model/labels/labels_stringlabels.go

```diff
@@ -425,34 +425,50 @@ func FromStrings(ss ...string) Labels {
 // TODO: replace with Less function - Compare is never needed.
 // TODO: just compare the underlying strings when we don't need alphanumeric sorting.
 func Compare(a, b Labels) int {
-	l := len(a.data)
-	if len(b.data) < l {
-		l = len(b.data)
-	}
-
-	ia, ib := 0, 0
-	for ia < l {
-		var aName, bName string
-		aName, ia = decodeString(a.data, ia)
-		bName, ib = decodeString(b.data, ib)
-		if aName != bName {
-			if aName < bName {
-				return -1
-			}
-			return 1
-		}
-		var aValue, bValue string
-		aValue, ia = decodeString(a.data, ia)
-		bValue, ib = decodeString(b.data, ib)
-		if aValue != bValue {
-			if aValue < bValue {
-				return -1
-			}
-			return 1
+	// Find the first byte in the string where a and b differ.
+	shorter, longer := a.data, b.data
+	if len(b.data) < len(a.data) {
+		shorter, longer = b.data, a.data
+	}
+	if len(shorter) == 0 {
+		return len(longer)
+	}
+	i := 0
+	_ = longer[len(shorter)-1] // Get compiler to do bounds-check on longer just once here.
+	for ; i < len(shorter); i++ {
+		if shorter[i] != longer[i] {
+			break
+		}
+	}
```

Comment:
Since `decodeSize` is very cheap, we can keep decoding sizes as we go, like:
```suggestion
	i := 0
	lastLblStart, nextLblStart := 0, 0
	_ = longer[len(shorter)-1] // Get compiler to do bounds-check on longer just once here.
	for ; i < len(shorter); i++ {
		if shorter[i] != longer[i] {
			break
		}
		if i >= nextLblStart {
			size, nextI := decodeSize(shorter, i)
			nextLblStart = nextI + size
			lastLblStart = i
		}
	}
```

So then later we can just do `i = lastLblStart`.

This gives pretty good performance impact:
```
goos: linux
goarch: amd64
pkg: github.com/prometheus/prometheus/model/labels
cpu: 11th Gen Intel(R) Core(TM) i7-11700K @ 3.60GHz
                                  │ bryanagain  │              tracknoloop               │
                                  │   sec/op    │   sec/op     vs base                   │
Labels_Compare/equal-16             25.59n ± 0%   26.48n ± 3%   +3.46% (p=0.000 n=30+10)
Labels_Compare/not_equal-16         30.25n ± 1%   24.07n ± 3%  -20.43% (p=0.000 n=30+10)
Labels_Compare/different_sizes-16   16.46n ± 1%   12.38n ± 8%  -24.84% (p=0.000 n=30+10)
Labels_Compare/lots-16              47.77n ± 1%   42.97n ± 4%  -10.06% (p=0.000 n=30+10)
geomean                             27.94n        24.13n       -13.63%

```

I've ran this benchmark multiple times and didn't get a consistent result (I was getting the `equals` benchmark faster in my version, which doesn't make sense), however I also think that if we do this, we can remove the last for loop (as we already know which labels are different, no need to search for them).

Reply:
In other words, I think we can do this later:
```go
	aStr, _ := decodeString(a.data, lastLblStart)
	bStr, _ := decodeString(b.data, lastLblStart)
```

Reply:
I would like to search for the first difference much faster, e.g. 8 bytes at a time using dwords, or even longer strides using SIMD.  I made various attempts at this, e.g. 06f8bc7787435a93626c96c3f0ea53b724ef23b4.  I think it's faster than your version, although I don't really like the implementation in `github.com/grailbio/base/simd.FirstUnequal8Unsafe`.

Putting an extra if-statement inside the tight loop does not seem attractive.

Reply:
I included your latest code, and updated my benchmark results.
Now mostly going faster than the slice version.